### ISSUE TITLE
scripts within imported template are not executed correctly in IE and FF

### ISSUE THREAD

**Issue by tomalec** [Juicy/imported-template#17]
This is the issue that was partially worked-around by https://github.com/Juicy/imported-template/commit/74a96309ba2d93712706cc5cca8eccf8da06cc46 and few more.
I post it, as this still makes problems, so it's better to describe and investigate the problem more.

It seems to be cause by an issue with HTML Imports/Tempalte polyfill - https://github.com/webcomponents/webcomponentsjs/issues/470, as it happens only in polyfilled environment.

**Comment by tomalec** [comment]
I have removed HTML Imports polyfill out of question, it seems to be template + separate document issue.
@miyconst @waprech do you have IE Edge to test if it works with IE's native HTMLTemplateElement implementation?

**Comment by tomalec** [comment]
@miyconst Could you check if it works with all you use cases? and maybe add some automated tests?

**Comment by miyconst** [comment]
I just have to take latest `imported-template` and check it, is that correct?

**Comment by tomalec** [comment]
Yes.
4 sty 2016 20:34 "Konstantin" <notifications@github.com> napisał(a):

**Comment by miyconst** [comment]
@tomalec here is the testing result:

- The scripts are executed in Chrome, IE, Edge, Safari, FF.
- The styles are applied in Chrome, IE, Edge, Safari, FF.
- In Safari the script loaded with `imported-template` being executed more than once.
- In IE & FF `document.currentScript.previousElementSibling` concept does not work. It looks like the `previousElementSibling` on the time of the script execution is not the `<template dom-bind>` which is actual `previousElementSibling`.

Here are the pages I tested with:

### index.html

```html
<!DOCTYPE html>
<html>
<head>
    <title></title>
	<meta charset="utf-8" />
    <script src="../webcomponentsjs/webcomponents.js"></script>
    <link href="../imported-template/imported-template.html" rel="import" />
    <link href="../polymer/polymer.html" rel="import" />
</head>
<body>
    <template is="dom-bind">
        <div>index.html</div>
        <template is="imported-template" content="/playground/sub.html"></template>
    </template>
    <script>
        console.log("In index.html!");
    </script>
</body>
</html>
```

### sub.html

```html
<template>
    <template is="dom-bind">
        <div>sub.html <span>{{model.value}}</span> <button on-click="plus">++</button></div>
    </template>
    <script>
        console.log("In sub.html!");
        (function () {
            var script = document._currentScript || document.currentScript;
            var template = script.previousElementSibling;

            template.model = {
                value: 1
            };

            template.plus = function () {
                template.set("model.value", template.model.value + 1);
            };
        })();
    </script>
</template>
```

**Comment by tomalec** [comment]
I have added a (quite ugly) workaround that should help with `previousElementSibling`. If it helps, adn we would like not to mess with polyfills methods, we could duplicate their code instead of [using](https://github.com/Juicy/imported-template/blob/10b93469e7d5233cc03334bd4490114772d25060/imported-template.html#L93) `HTMLImports.parser`

I will check how it works with `dom-bind` tomorrow

It's hard for me to debug Safari issue, could you give some insight?

**Comment by miyconst** [comment]
I don't really understand what this monkey patch does. The `previousElementSibling` does not work with or without it in IE. And this workaround works with or without the patch.

```js
var script = document._currentScript || document.currentScript;
var template = script.previousElementSibling;

if (template.getAttribute("is") != "dom-bind") {
	template = template.previousElementSibling;
}
```

This workaround also works in FF.

**Comment by miyconst** [comment]
And it also works inside `dom-bind` with/without `dom-repeat`.

**Comment by miyconst** [comment]
>It's hard for me to debug Safari issue, could you give some insight?

It looks like Safari executes script twice, first time when the content was added, same as Chrome, and second time when `parseDynamic` is invoked.

https://github.com/Juicy/imported-template/blob/gh-pages/imported-template.html#L107

The first time value of `script.previousElementSibling` is correct, the second time `script.previousElementSibling.previousElementSibling` should be used, as expected.

**Comment by miyconst** [comment]
I have added a check whether it's Safari, and don't dynamic parse content if so.

**Comment by tomalec** [comment]
Thanks for Safari fix.
I have add another fix for my hack, could you check once again if `script.previousElementSibling` gives correct node without additional check in your code?

**Comment by miyconst** [comment]
Your hack works inside `dom-bind` and works for the first element inside `dom-pereat`, but does not work for the second and other elements inside `dom-repeat`.

**Comment by miyconst** [comment]
Here is my page with `dom-repeat`.

```html
<!DOCTYPE html>
<html>
<head>
    <title></title>
	<meta charset="utf-8" />
    <script src="../webcomponentsjs/webcomponents.js"></script>
    <link href="../imported-template/imported-template.html" rel="import" />
    <link href="../polymer/polymer.html" rel="import" />
</head>
<body>
    <template is="dom-bind">
        <div>index.html</div>
        <template is="dom-repeat" items="{{model.items}}">
            <fieldset>
                <legend>Item <span>{{item}}</span></legend>
                <template is="imported-template" content="/playground/sub.html"></template>
            </fieldset>
        </template>
    </template>
    <script>
        (function () {
            var script = document._currentScript || document.currentScript;
            var template = script.previousElementSibling;

            //if (template.getAttribute("is") != "dom-bind") {
            //    template = template.previousElementSibling;
            //}

            template.model = {
                items: [1, 2, 3]
            };
        })();
        console.log("In index.html!");
    </script>
</body>
</html>
```

And the sub.html

```html
<template>
    <template is="dom-bind" id="sub-dom-bind">
        <div class="sub">sub.html <span>{{model.value}}</span> <button on-click="plus">++</button></div>
        <!--<template is="imported-template" content="/playground/deep.html"></template>-->
    </template>
    <script id="sub-script">
        (function () {
            var script = document._currentScript || document.currentScript;
            var template = script.previousElementSibling;

            //if (template.getAttribute("is") != "dom-bind") {
            //    template = template.previousElementSibling;
            //}

            template.model = {
                value: 1
            };

            template.plus = function () {
                template.set("model.value", template.model.value + 1);
            };

            console.log(script.id, template.id, template.getAttribute("is"));
        })();
        console.log("In sub.html!");
    </script>
    <style>
        div.sub {
            font-weight: bold;
            color: red;
        }
    </style>
</template>
```

**Comment by tomalec** [comment]
@miyconst I have worked around this as well, I hope this is the last hack we need here.

I simplified and automated your use cases here https://github.com/Juicy/imported-template/tree/gh-pages/test/use-cases, [all tests are passing](https://travis-ci.org/Juicy/imported-template) every where except Safari 9 - which is hard to track for me w/o a Mac

Do you have any more failing use cases?

**Comment by miyconst** [comment]
Great job @tomalec, once again!

I have tested deep nesting, with/without `dom-bind`, `dom-repeat`, `dom-if`, everything works in Chrome, Edge (with native template support and without it), IE, FF, Safari.

>except Safari 9 - which is hard to track for me w/o a Mac

How do I test it? For me all the tests pass.

Edge with native template support (`v25.10586.0.0`) does not work without those hacks:
https://github.com/Juicy/imported-template/blob/gh-pages/imported-template.html#L112
https://github.com/Juicy/imported-template/blob/gh-pages/imported-template.html#L122

**Comment by tomalec** [comment]
Thanks,
I have separated Safari 9 issue to https://github.com/Juicy/imported-template/issues/19

> Edge with native template support (v25.10586.0.0) does not work without those hacks
It would be cool to investigate it, and maybe there is an issue in Chackra to be reported

**Comment by tomalec** [comment]
Hopefully that's  the end of hacks on wc.js.
Released with [1.1.0](https://github.com/Juicy/imported-template/releases/tag/1.1.0)
